### PR TITLE
Fixes #415 - Close stdin when tearing down subprocesses in system-tests to prevent fd leaks

### LINKED ISSUE #415: ResourceWarning: unclosed file <_io.BufferedWriter name=12>
https://github.com/skupperproject/skupper-router/runs/6243241095?check_suite_focus=true#step:9:1043

```
18: ::OneRouterTest::test_01_listen_error
18:   /usr/lib/python3.8/unittest/case.py:633: ResourceWarning: unclosed file <_io.BufferedWriter name=12>
18:     method()
18:   
18:   Object allocated at:
18:     File "/home/runner/.local/lib/python3.8/site-packages/pluggy/_manager.py", line 80
18:       return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
18:     File "/home/runner/.local/lib/python3.8/site-packages/pluggy/_callers.py", line 39
18:       res = hook_impl.function(*args)
18:     File "/home/runner/.local/lib/python3.8/site-packages/_pytest/main.py", line 347
18:       item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
18:     File "/home/runner/.local/lib/python3.8/site-packages/pluggy/_hooks.py", line 265
18:       return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
18:     File "/home/runner/.local/lib/python3.8/site-packages/pluggy/_manager.py", line 80
18:       return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
18:     File "/home/runner/.local/lib/python3.8/site-packages/pluggy/_callers.py", line 39
18:       res = hook_impl.function(*args)
18:     File "/home/runner/.local/lib/python3.8/site-packages/_pytest/runner.py", line 111
18:       runtestprotocol(item, nextitem=nextitem)
18:     File "/home/runner/.local/lib/python3.8/site-packages/_pytest/runner.py", line 130
18:       reports.append(call_and_report(item, "call", log))
18:     File "/home/runner/.local/lib/python3.8/site-packages/_pytest/runner.py", line 219
18:       call = call_runtest_hook(item, when, **kwds)
18:     File "/home/runner/.local/lib/python3.8/site-packages/_pytest/runner.py", line 258
18:       return CallInfo.from_call(
18:     File "/home/runner/.local/lib/python3.8/site-packages/_pytest/runner.py", line 338
18:       result: Optional[TResult] = func()
18:     File "/home/runner/.local/lib/python3.8/site-packages/_pytest/runner.py", line 259
18:       lambda: ihook(item=item, **kwds), when=when, reraise=reraise
18:     File "/home/runner/.local/lib/python3.8/site-packages/pluggy/_hooks.py", line 265
18:       return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
18:     File "/home/runner/.local/lib/python3.8/site-packages/pluggy/_manager.py", line 80
18:       return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
18:     File "/home/runner/.local/lib/python3.8/site-packages/pluggy/_callers.py", line 39
18:       res = hook_impl.function(*args)
18:     File "/home/runner/.local/lib/python3.8/site-packages/_pytest/runner.py", line 166
18:       item.runtest()
18:     File "/home/runner/.local/lib/python3.8/site-packages/_pytest/unittest.py", line 327
18:       self._testcase(result=self)  # type: ignore[arg-type]
18:     File "/usr/lib/python3.8/unittest/case.py", line 736
18:       return self.run(*args, **kwds)
18:     File "/usr/lib/python3.8/unittest/case.py", line 676
18:       self._callTestMethod(testMethod)
18:     File "/usr/lib/python3.8/unittest/case.py", line 633
18:       method()
```

The stacktrace above is useless, but the test itself is simple and there appears only one place which deals with fd's, the subprocess call.

https://github.com/skupperproject/skupper-router/blob/4dac88768799142f50f64f086a707c6bbe66e9e0/tests/system_tests_one_router.py#L409-L415

Looking at the code, it looks sufficient to me. The code calls `subprocess.wait`, which I always thought is enough. And, trying with `/usr/bin/python-debug`, it appears it is enough.  Unless the subprocess redirects outputs, but qdrouterd does not do anything like the following

```
import subprocess

p = subprocess.Popen(["bash", "-c", "echo baf"], stdout=subprocess.PIPE)
p.wait()

# produces
# $ python-debug -Xdev scratches/scratch_8.py
# sys:1: ResourceWarning: unclosed file <_io.BufferedReader name=3>
```

But `Process` does!

https://github.com/skupperproject/skupper-router/blob/fdfd076add2e565f062e9e7e63db22859c7b534e/tests/system_test.py#L300-L310

I am surprised one can get away with the `with` since the fd clearly escapes out of the `with`'s scope...

The `subprocess.Popen#__exit__` in the standard library is implemented like this

```python
    def __exit__(self, exc_type, value, traceback):
        if self.stdout:
            self.stdout.close()
        if self.stderr:
            self.stderr.close()
        try:  # Flushing a BufferedWriter may raise an error
            if self.stdin:
                self.stdin.close()
        finally:
            if exc_type == KeyboardInterrupt:
                # https://bugs.python.org/issue25942
                # In the case of a KeyboardInterrupt we assume the SIGINT
                # was also already sent to our child processes.  We can't
                # block indefinitely as that is not user friendly.
                # If we have not already waited a brief amount of time in
                # an interrupted .wait() or .communicate() call, do so here
                # for consistency.
                if self._sigint_wait_secs > 0:
                    try:
                        self._wait(timeout=self._sigint_wait_secs)
                    except TimeoutExpired:
                        pass
                self._sigint_wait_secs = 0  # Note that this has been done.
                return  # resume the KeyboardInterrupt

            # Wait for the process to terminate, to avoid zombies.
            self.wait()
```

### PR DESCRIPTION
**Pull request by jiridanek**

### CODE REVIEW

File: tests/system_test.py [hunk 1]
```diff
@@ -372,6 +372,10 @@ def error(msg):
         # at this point the process has terminated, either of its own accord or
         # due to the above terminate call.
 
+        # prevent leak of stdin fd
```
Comment (ganeshmurthy) [inline tests/system_test.py:375]: Is closing the stdin the only fix needed in order not to see the backtrace seen in https://github.com/skupperproject/skupper-router/issues/415 ?

Comment (jiridanek) [reply tests/system_test.py:375]: Yes, close stdin and ensure that the `teardown` method is always called.

Comment (ganeshmurthy) [reply tests/system_test.py:375]: Ok, so you added the `__enter__` and `__exit__` functions to `class Qdrouterd` so if anyone uses the `with` keyword, we can ensure that the `teardown()` function will be called. But adding these two functions and the changes to the other tests in this commit are not part of the fix. The *actual* fix is to close the stdin on `teardown()`, right ?

Comment (jiridanek) [reply tests/system_test.py:375]: It is part of the fix. I need to ensure that stain gets closed. If some tests never called teardown() and I am closing stdin in teardown, then I have to also call teardown there.

Am I missing something obvious here?

Comment (jiridanek) [reply tests/system_test.py:375]: I think I get it now. The change in system_tests_management.py from assertTrue to assertEqual is unrelated, yes. Let me put it to separate PR.

Comment (jiridanek) [reply tests/system_test.py:375]: Split into https://github.com/skupperproject/skupper-router/pull/508, thanks!

**Review by ganeshmurthy — APPROVED** [review]

**Pull request merged**